File: apkEditor/src/main/java/com/mcal/apkeditor/dialogs/PatchDialog.java
package com.mcal.apkeditor.dialogs;

import android.content.res.AssetManager;
import android.graphics.Color;
import android.graphics.Typeface;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.webkit.WebView;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mcal.apkeditor.activities.ApkInfoActivity;
import com.mcal.apkeditor.R;
import com.mcal.apkeditor.ce.ManifestInfo;
import com.mcal.apkeditor.ce.ManifestParser;
import com.mcal.apkeditor.dialogs.FileSelectDialog.IFileSelection;
import com.mcal.apkeditor.patch.IPatchContext;
import com.mcal.apkeditor.patch.PatchExecutor;
import com.mcal.apkeditor.view.ViewDialog;
import com.mcal.common.utils.IOUtils;
import com.mcal.common.utils.SDCard;

import java.io.Closeable;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

// Dialog used for patch applying
public class PatchDialog
        implements android.view.View.OnClickListener, IPatchContext {

    private final WeakReference<ApkInfoActivity> activityRef;
    // Record all the global parameter values
    private final Map<String, String> globalVariableValues = new HashMap<String, String>();
    private String exampleDir = null;
    private String patchPath = null;
    private TextView curPatchTv;
    private TextView patchPathTv;
    private TextView saveExamplesTv;
    private Button selectApplyBtn;
    private WebView webView;
    private View logLayout;
    private TextView logTv;
    // Manifest info parsed from original apk
    private ManifestInfo manifestInfo;
    // Record executor as the parse is done there
    private PatchExecutor patchExecutor;
    private ViewDialog dialog;

    public PatchDialog(ApkInfoActivity activity) {

        this.activityRef = new WeakReference<>(activity);

        init(activity);
    }

    private void init(ApkInfoActivity activity) {

        View view = LayoutInflater.from(activity).inflate(R.layout.dlg_patch, null);

        this.curPatchTv = view.findViewById(R.id.tv_curpatch);
        curPatchTv.setOnClickListener(this);
        this.patchPathTv = view.findViewById(R.id.tv_patch_path);
        Button closeBtn = view.findViewById(R.id.btn_close);
        closeBtn.setOnClickListener(this);
        this.selectApplyBtn = view.findViewById(R.id.btn_select_apply);
        selectApplyBtn.setOnClickListener(this);
        this.saveExamplesTv = view
                .findViewById(R.id.tv_save_patches);
        saveExamplesTv.setOnClickListener(this);

        this.webView = view.findViewById(R.id.web_instructions);
        webView.loadUrl("file:///android_res/raw/about_patch.html");
        this.logLayout = view.findViewById(R.id.log_layout);
        this.logTv = view.findViewById(R.id.tv_patchlog);

        dialog = new ViewDialog(activity);
        dialog.setTitle(R.string.patch_capital);
        dialog.setView(view);
    }

    public void show() {
        dialog.show();
    }

    @Override
    public void onClick(@NonNull View v) {
        int id = v.getId();
        if (id == R.id.btn_close) {
            dialog.dismiss();
        }
        // Click on current patch
        else if (id == R.id.tv_curpatch) {
            selectPatch();
        }
        // Select a patch or apply it
        else if (id == R.id.btn_select_apply) {
            if (patchPath == null) {
                selectPatch();
            } else {
                applyPatch();
            }
        }
        // Extract and save the examples
        else if (id == R.id.tv_save_patches) {
            boolean ret = extractExamples("patch_app_rename.zip");
            ret |= extractExamples("patch_data_editor.zip");
            ret |= extractExamples("patch_new_entrance.zip");
            ret |= extractExamples("patch_my_font.zip");
            ret |= extractExamples("patch_mem_editor.zip");
            ret |= extractExamples("patch_bypass_sigcheck.zip");
            ret |= extractExamples("patch_launcher_toast.zip");
            ret |= extractExamples("patch_script_example.zip");
            if (ret) {
                String message = String.format(
                        activityRef.get()
                                .getString(R.string.patch_examples_copied),
                        this.exampleDir);
                Toast.makeText(activityRef.get(), message, Toast.LENGTH_SHORT)
                        .show();
            }
        }
    }

    private void initExampleDir() {
        // Check the directory exist or not
        if (this.exampleDir == null) {
            try {
                exampleDir = SDCard.makeDir(activityRef.get(), "examples");
            } catch (Exception e1) {
                e1.printStackTrace();
            }
        }
    }

    private boolean extractExamples(String filename) {
        initExampleDir();

        String path = exampleDir + filename;

        AssetManager am = activityRef.get().getAssets();
        InputStream input = null;
        FileOutputStream output = null;
        try {
            input = am.open(filename);
            output = new FileOutputStream(path);
            IOUtils.copy(input, output);

            return true;
        } catch (IOException e) {
            Toast.makeText(activityRef.get(), e.getMessage(),
                    Toast.LENGTH_SHORT).show();
        } finally {
            closeQuietly(input);
            closeQuietly(output);
        }

        return false;
    }

    private void closeQuietly(Closeable c) {
        if (c != null) {
            try {
                c.close();
            } catch (IOException ignored) {
            }
        }
    }

    private void closeQuietly(ZipFile zfile) {
        if (zfile != null) {
            try {
                zfile.close();
            } catch (IOException ignored) {
            }
        }
    }

    private void applyPatch() {
        // Switch the view
        logTv.setText("");
        selectApplyBtn.setEnabled(false);

        // Patch it
        patchExecutor = new PatchExecutor(activityRef.get(), patchPath, this);
        patchExecutor.applyPatch();
    }

    private void selectPatch() {
        String defaultDir = null;
        initExampleDir();
        if (new File(exampleDir).exists()) {
            defaultDir = exampleDir;
        }
        new FileSelectDialog(activityRef.get(),
                new IFileSelection() {
                    @Override
                    public void fileSelectedInDialog(
                            String filePath, String extraStr, boolean openFile) {
                        patchSelected(filePath);
                    }

                    @Override
                    public boolean isInterestedFile(String filename, String extraStr) {
                        return filename.endsWith(".zip");
                    }

                    @Override
                    public String getConfirmMessage(String filePath,
                                                    String extraStr) {
                        return null;
                    }
                }, ".zip", null,
                activityRef.get().getString(R.string.select_patch), false, false, false, "patch", defaultDir);
    }

    private void patchSelected(String filePath) {
        logLayout.setVisibility(View.VISIBLE);
        webView.setVisibility(View.GONE);

        String patchConfig = getPatchConfig(filePath);
        if (patchConfig != null) {
            logTv.setText(patchConfig);
        }

        this.patchPath = filePath;
        this.patchPathTv.setText(patchPath);
        this.selectApplyBtn.setText(R.string.apply_the_patch);
    }

    @Nullable
    private String getPatchConfig(String filePath) {
        ZipFile zfile = null;
        InputStream input = null;
        try {
            zfile = new ZipFile(filePath);
            ZipEntry entry = zfile.getEntry("patch.txt");
            if (entry == null) {
                this.error(R.string.patch_error_no_entry, "patch.txt");
            }

            input = zfile.getInputStream(entry);
            return IOUtils.readString(input);
        } catch (Exception e) {
            error(R.string.general_error, e.getMessage());
        } finally {
            closeQuietly(input);
            closeQuietly(zfile);
        }
        return null;
    }

    @Override
    public void info(int resourceId, boolean bold, Object... args) {
        String txt = activityRef.get().getString(resourceId);
        if (args != null) {
            txt = String.format(txt, args);
        }
        String message = bold ? ("\n" + txt + "\n") : (txt + "\n");
        appendText(message, bold, false);
    }

    @Override
    public void info(String format, boolean bold, Object... args) {
        String txt = format;
        if (args != null) {
            txt = String.format(txt, args);
        }
        String message = bold ? ("\n" + txt + "\n") : (txt + "\n");
        appendText(message, bold, false);
    }

    @Override
    public void error(int resourceId, Object... args) {
        String txt = activityRef.get().getString(resourceId);
        if (args != null) {
            txt = String.format(txt, args);
        }
        appendText(txt + "\n", false, true);
    }

    private void appendText(final String txt, final boolean bold,
                            final boolean red) {
        activityRef.get().runOnUiThread(() -> {
            if (red) {
                SpannableString spanString = new SpannableString(txt);
                ForegroundColorSpan span = new ForegroundColorSpan(
                        Color.RED);
                spanString.setSpan(span, 0, txt.length(),
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                logTv.append(spanString);
            } else if (bold) {
                SpannableString spanString = new SpannableString(txt);
                StyleSpan span = new StyleSpan(Typeface.BOLD);
                spanString.setSpan(span, 0, txt.length(),
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                logTv.append(spanString);
            } else {
                logTv.append(txt);
            }
        });
    }

    @Override
    public String getString(int stringId) {
        return activityRef.get().getString(stringId);
    }

    @Override
    public void setVariableValue(String key, String value) {
        globalVariableValues.put(key, value);
    }

    @Override
    public String getVariableValue(String key) {
        return globalVariableValues.get(key);
    }

    @Override
    public void patchFinished() {
        activityRef.get().runOnUiThread(() -> {
            selectApplyBtn.setText(R.string.patch_applied);
            selectApplyBtn.setBackgroundColor(0xff606060);
        });
    }

    @Override
    public String getDecodeRootPath() {
        return activityRef.get().getDecodeRootPath();
    }

    @Override
    public List<String> getSmaliFolders() {
        List<String> folders = new ArrayList<>();
        folders.add("smali");

        // Look into zip file to get all dex, thus get related smali folder
        String apkPath = activityRef.get().getApkPath();
        ZipFile zfile = null;
        try {
            zfile = new ZipFile(apkPath);
            Enumeration<? extends ZipEntry> entries = zfile.entries();
            while (entries.hasMoreElements()) {
                ZipEntry entry = entries.nextElement();
                String name = entry.getName();
                if (name.endsWith(".dex") && !name.contains("/")) {
                    if (!name.equals("classes.dex")) {
                        folders.add("smali_"
                                + name.substring(0, name.length() - 4));
                    }
                }
            }
        } catch (IOException e1) {
            e1.printStackTrace();
        } finally {
            closeQuietly(zfile);
        }

        return folders;
    }

    private void initManifestInfo() {
        String apkPath = activityRef.get().getApkPath();
        ZipFile zfile = null;
        InputStream input = null;
        try {
            zfile = new ZipFile(apkPath);
            ZipEntry zipEntry = zfile.getEntry("AndroidManifest.xml");
            input = zfile.getInputStream(zipEntry);
            ManifestParser parser = new com.mcal.apkeditor.ce.ManifestParser(
                    input);
            parser.parse();
            this.manifestInfo = parser.getManifestInfo();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            closeQuietly(input);
            closeQuietly(zfile);
        }
    }

    @Override
    public String getApplication() {
        if (this.manifestInfo == null) {
            initManifestInfo();
        }
        if (this.manifestInfo != null) {
            String name = manifestInfo.applicationCls;
            if (name == null) {
                return null;
            }
            if (name.startsWith(".")) {
                name = manifestInfo.packageName + name;
            } else if (!name.contains(".")) {
                name = manifestInfo.packageName + "." + name;
            }
            return name;
        }
        return null;
    }

    @Override
    public List<String> getActivities() {
        if (this.manifestInfo == null) {
            initManifestInfo();
        }
        if (this.manifestInfo != null) {
            List<String> activities = new ArrayList<>();
            for (Entry<Integer, Integer> entry : manifestInfo.compNameIdx2Type.entrySet()) {
                // it is activity
                if (entry.getValue() == 0) {
                    String name = getComponentName(manifestInfo, entry.getKey());
                    if (!activities.contains(name)) {
                        activities.add(name);
                    }
                }
            }
            return activities;
        }

        return null;
    }

    @Override
    public List<String> getLauncherActivities() {
        if (this.manifestInfo == null) {
            initManifestInfo();
        }
        if (this.manifestInfo != null) {
            List<String> activities = new ArrayList<>();
            for (Integer idx : manifestInfo.launcherActIdxs) {
                String name = getComponentName(manifestInfo, idx);
                if (!activities.contains(name)) {
                    activities.add(name);
                }
            }
            return activities;
        }

        return null;
    }

    // Get real component name
    @Nullable
    private String getComponentName(@NonNull ManifestInfo manifestInfo, int index) {
        if (manifestInfo.targetActivityIdxs != null
                && manifestInfo.targetActivityIdxs.containsKey(index)) {
            index = manifestInfo.targetActivityIdxs.get(index);
        }

        if (index < manifestInfo.strings.size()) {
            String name = manifestInfo.strings.get(index);
            if (name.startsWith(".")) {
                name = manifestInfo.packageName + name;
            } else if (!name.contains(".")) {
                name = manifestInfo.packageName + "." + name;
            }
            return name;
        }

        return null;
    }

    @Override
    public List<String> getPatchNames() {
        if (patchExecutor != null) {
            return patchExecutor.getRuleNames();
        }
        return null;
    }
}

File: apkEditor/src/main/java/com/mcal/apkeditor/dialogs/PatchDialog.kt
package com.mcal.apkeditor.dialogs

import android.app.Activity
import android.content.DialogInterface
import android.graphics.Color
import android.graphics.Typeface
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.webkit.WebView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.mcal.apkeditor.R
import com.mcal.apkeditor.dialogs.FileSelectDialog.IFileSelection
import com.mcal.apkeditor.patch.PatchExecutor
import com.mcal.apkeditor.patch.interfaces.ApkInfoListener
import com.mcal.apkeditor.patch.interfaces.IPatchContext
import com.mcal.common.utils.LocaleManager
import com.mcal.common.utils.copyFile
import com.mcal.common.utils.makeDir
import com.mcal.common.utils.readText
import com.mcal.patchview.ui.CodeText
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.xml.sax.SAXException
import ru.mcal.manifestparser.xml.AndroidManifestParser
import java.io.*
import java.util.zip.ZipFile
import javax.xml.parsers.ParserConfigurationException

// Dialog used for patch applying
class PatchDialog(activity: Activity, private val listener: ApkInfoListener) : View.OnClickListener,
    IPatchContext {
    private val mActivity = activity

    // Record all the global parameter values
    private val globalVariableValues: MutableMap<String, String> = HashMap()
    private var exampleDir: String? = null
    private var patchPath: String? = null
    private var patchPathTv: TextView? = null
    private var webView: WebView? = null
    private var logLayout: View? = null
    private var logTv: CodeText? = null

    // Record executor as the parse is done there
    private var patchExecutor: PatchExecutor? = null
    private var materialDialog: AlertDialog? = null
    private val manifestPath: File = File(listener.decodeRootPath + "/AndroidManifest.xml")

    private fun init(activity: Activity) {
        val view = LayoutInflater.from(activity).inflate(R.layout.dialog_patch, null)
        val curPatchTv = view.findViewById<TextView>(R.id.tv_curpatch)
        curPatchTv.setOnClickListener(this)
        patchPathTv = view.findViewById(R.id.tv_patch_path)
        val saveExamplesTv = view.findViewById<TextView>(R.id.tv_save_patches)
        saveExamplesTv.setOnClickListener(this)
        webView = view.findViewById(R.id.web_instructions)
        CoroutineScope(Dispatchers.Main).launch {
            webView?.loadUrl("file:///android_asset/doc/" + LocaleManager.getDocLanguage() + "/patch.html")
        }
        logLayout = view.findViewById(R.id.log_layout)
        logTv = view.findViewById(R.id.tv_patchlog)
        materialDialog = MaterialAlertDialogBuilder(activity)
            .setView(view)
            .setPositiveButton(activity.getString(R.string.select_patch), null)
            .setNegativeButton(activity.getString(R.string.close), null)
            .create()
        materialDialog?.setOnShowListener {
            materialDialog?.getButton(DialogInterface.BUTTON_POSITIVE)?.setOnClickListener {
                if (patchPath == null) {
                    selectPatch()
                } else {
                    applyPatch()
                }
            }
        }
        materialDialog?.show()
    }

    override fun onClick(v: View) {
        val id = v.id
        if (id == R.id.btn_close) {
            materialDialog?.dismiss()
        } else if (id == R.id.tv_curpatch) {
            selectPatch()
        } else if (id == R.id.tv_save_patches) {
            // ApkEditor Patches
            var ret = extractExamples("patch_AE_app_rename.zip")
            ret = ret or extractExamples("patch_AE_bypass_sign_check.zip")
            ret = ret or extractExamples("patch_AE_data_editor.zip")
            ret = ret or extractExamples("patch_AE_launcher_toast.zip")
            ret = ret or extractExamples("patch_AE_mem_editor.zip")
            ret = ret or extractExamples("patch_AE_my_font.zip")
            ret = ret or extractExamples("patch_AE_new_entrance.zip")
            ret = ret or extractExamples("patch_AE_script_example.zip")

            // Kubarev Patches
            ret = ret or extractExamples("patch_Bin_SignHook.zip")
            ret = ret or extractExamples("patch_CNFIX_3.0_SignHook.zip")
            ret = ret or extractExamples("patch_Heavenly_SignHook.zip")
            ret = ret or extractExamples("patch_LP_DexSignHook.zip")
            ret = ret or extractExamples("patch_LP_SignHook.zip")
            ret = ret or extractExamples("patch_R3Tools_SignHook.zip")
            ret = ret or extractExamples("patch_Ultima_SignHook.zip")
            ret = ret or extractExamples("patch_Ultima_VipSignHook.zip")
            if (ret) {
                val message =
                    String.format(mActivity.getString(R.string.patch_examples_copied), exampleDir)
                Toast.makeText(mActivity, message, Toast.LENGTH_SHORT)
                    .show()
            }

//            mActivity.assets.list("a")?.forEach { patch ->
//                extractExamples(patch)
//            }

        }
    }

    private fun initExampleDir() {
        // Check the directory exist or not
        if (exampleDir == null) {
            try {
                exampleDir = makeDir("patches").path
            } catch (e1: Exception) {
                e1.printStackTrace()
            }
        }
    }

    private fun extractExamples(filename: String): Boolean {
        initExampleDir()
        val path = exampleDir + File.separator + filename
        val am = mActivity.assets
        var input: InputStream? = null
        var output: FileOutputStream? = null
        try {
            input = am.open("patches" + File.separator + filename)
            output = FileOutputStream(path)
            copyFile(input, output)
            return true
        } catch (e: IOException) {
            Toast.makeText(
                this.mActivity, e.message,
                Toast.LENGTH_SHORT
            ).show()
        } finally {
            closeQuietly(input)
            closeQuietly(output)
        }
        return false
    }

    private fun closeQuietly(close: Closeable?) {
        close?.let {
            try {
                close.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    private fun closeQuietly(zfile: ZipFile?) {
        zfile?.let {
            try {
                zfile.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    private fun applyPatch() {
        // Switch the view
        logTv?.text = ""

        materialDialog?.getButton(DialogInterface.BUTTON_POSITIVE)?.isEnabled = false
        materialDialog?.show()

        // Patch it
        // TODO: Multiple select patches
        patchExecutor = PatchExecutor(mActivity, listener, patchPath, this)
        patchExecutor?.applyPatch()
    }

    private fun selectPatch() {
        var defaultDir: String? = null
        initExampleDir()
        exampleDir?.let { dir ->
            if (File(dir).exists()) {
                defaultDir = exampleDir
            }
        }
        FileSelectDialog(
            mActivity,
            object : IFileSelection {
                override fun fileSelectedInDialog(
                    filePath: String?, extraStr: String?, openFile: Boolean
                ) {
                    filePath?.let {
                        patchSelected(filePath)
                    }
                }

                override fun isInterestedFile(filename: String?, extraStr: String?): Boolean {
                    filename?.let {
                        return filename.endsWith(".zip")
                    }
                    return false
                }

                override fun getConfirmMessage(filePath: String?, extraStr: String?): String? {
                    return null
                }

            },
            ".zip",
            null,
            mActivity.getString(R.string.select_patch),
            false,
            false,
            false,
            "patch",
            defaultDir
        )
    }

    private fun patchSelected(filePath: String) {
        logLayout?.visibility = View.VISIBLE
        webView?.visibility = View.GONE
        val patchConfig = getPatchConfig(filePath)
        patchConfig?.let { config ->
            logTv?.text = config
        }
        patchPath = filePath
        patchPathTv?.text = patchPath
        materialDialog?.getButton(DialogInterface.BUTTON_POSITIVE)
            ?.setText(R.string.apply_the_patch)
        materialDialog?.show()
    }

    private fun getPatchConfig(filePath: String): String? {
        var zfile: ZipFile? = null
        var input: InputStream? = null
        try {
            zfile = ZipFile(filePath)
            val entry = zfile.getEntry("patch.txt")
            if (entry == null) {
                this.error(R.string.patch_error_no_entry, "patch.txt")
            }
            input = zfile.getInputStream(entry)
            return input.readText()
        } catch (e: Exception) {
            e.message?.let { message ->
                error(R.string.general_error, message)
            }
        } finally {
            closeQuietly(input)
            closeQuietly(zfile)
        }
        return null
    }

    @Suppress("SAFE_CALL_WILL_CHANGE_NULLABILITY")
    override fun info(resourceId: Int, bold: Boolean, vararg args: Any) {
        var txt = mActivity.getString(resourceId)
        @Suppress("UNNECESSARY_SAFE_CALL")
        args?.let {
            txt = String.format(txt, *args)
        }
        val message = if (bold) "\n" + txt + "\n" else txt + "\n"
        appendText(message, bold, false)
    }

    @Suppress("SAFE_CALL_WILL_CHANGE_NULLABILITY")
    override fun info(format: String, bold: Boolean, vararg args: Any) {
        var txt = format
        @Suppress("UNNECESSARY_SAFE_CALL")
        args?.let {
            txt = String.format(txt, *args)
        }
        val message = if (bold) "\n" + txt + "\n" else txt + "\n"
        appendText(message, bold, false)
    }

    @Suppress("SAFE_CALL_WILL_CHANGE_NULLABILITY")
    override fun error(resourceId: Int, vararg args: Any) {
        var txt = mActivity.getString(resourceId)
        @Suppress("UNNECESSARY_SAFE_CALL")
        args?.let {
            txt = String.format(txt, *args)
        }
        appendText(txt + "\n", bold = false, red = true)
    }

    private fun appendText(
        txt: String, bold: Boolean,
        red: Boolean
    ) {
        mActivity.runOnUiThread {
            if (red) {
                val spanString = SpannableString(txt)
                val span = ForegroundColorSpan(Color.RED)
                spanString.setSpan(span, 0, txt.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                logTv?.append(spanString)
            } else if (bold) {
                val spanString = SpannableString(txt)
                val span = StyleSpan(Typeface.BOLD)
                spanString.setSpan(
                    span, 0, txt.length,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                )
                logTv?.append(spanString)
            } else {
                logTv?.append(txt)
            }
        }
    }

    override fun getString(stringId: Int): String {
        return mActivity.getString(stringId)
    }

    override fun setVariableValue(key: String, value: String) {
        globalVariableValues[key] = value
    }

    override fun getVariableValue(key: String): String? {
        globalVariableValues[key]?.let {
            return it
        }
        return null
    }

    override fun patchFinished() {
        mActivity.runOnUiThread {
            materialDialog?.getButton(DialogInterface.BUTTON_POSITIVE)
                ?.setText(R.string.patch_applied)
            materialDialog?.getButton(DialogInterface.BUTTON_POSITIVE)
                ?.setBackgroundColor(-0x9f9fa0)
            materialDialog?.show()
        }
    }

    override fun getDecodeRootPath(): String {
        return listener.decodeRootPath
    }

    override fun getSmaliFolders(): List<String> {
        val folders: MutableList<String> = ArrayList()
        folders.add("smali")

        // Look into zip file to get all dex, thus get related smali folder
        val apkPath = listener.apkPath
        var zfile: ZipFile? = null
        try {
            zfile = ZipFile(apkPath)
            val entries = zfile.entries()
            while (entries.hasMoreElements()) {
                val entry = entries.nextElement()
                val name = entry.name
                if (name.endsWith(".dex") && !name.contains("/")) {
                    if (name != "classes.dex") {
                        folders.add(
                            "smali_" + name.substring(0, name.length - 4)
                        )
                    }
                }
            }
        } catch (e1: IOException) {
            e1.printStackTrace()
        } finally {
            closeQuietly(zfile)
        }
        return folders
    }

    override fun getApplicationName(): String? {
        try {
            val parser = AndroidManifestParser.parse(FileInputStream(manifestPath))
            val name = parser.applicationName
            name?.let {
                return name
            }
        } catch (e: ParserConfigurationException) {
            e.printStackTrace()
        } catch (e: SAXException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return null
    }

    @Suppress("SAFE_CALL_WILL_CHANGE_NULLABILITY")
    override fun getActivities(): List<String>? {
        val activityName: MutableList<String> = ArrayList()
        try {
            val parser = AndroidManifestParser.parse(FileInputStream(manifestPath))
            for (activity in parser.activities) {
                activityName.add(activity.name)
            }
            @Suppress("UNNECESSARY_SAFE_CALL")
            activityName?.let {
                return activityName
            }
        } catch (e: ParserConfigurationException) {
            e.printStackTrace()
        } catch (e: SAXException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return null
    }

    @Suppress("SAFE_CALL_WILL_CHANGE_NULLABILITY")
    override fun getLauncherActivities(): List<String>? {
        val activityName: MutableList<String> = ArrayList()
        try {
            val parser = AndroidManifestParser.parse(FileInputStream(manifestPath))
            activityName.add(parser.launcherActivity.name)
            @Suppress("UNNECESSARY_SAFE_CALL")
            activityName?.let {
                return activityName
            }
        } catch (e: ParserConfigurationException) {
            e.printStackTrace()
        } catch (e: SAXException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return null
    }

    override fun getPatchNames(): List<String>? {
        patchExecutor?.let { executor ->
            return executor.ruleNames
        }
        return null
    }

    init {
        init(activity)
    }
}